#' @import sccore magrittr dplyr
#' @importFrom rlang is_empty
#' @importFrom stringr str_ends
#' @importFrom Biostrings readBStringSet
#' @importFrom stats setNames
#' @importFrom pbmcapply pbmcmapply
NULL

#' Get tss region sequence of target genes in regulaotry relationships
#'
#' @param gtf Gene transfer format, you can download it from http://www.ensembl.org/info/data/ftp/index.html
#' @param gene.use character, indicating target genes
#' @param upstream_length numeric, indicating upstream region length of TSS
#' @param downstream_length numeric, indicating downstream region length of TSS
#' @return data.frame, contains four columns are target gene, chr, start and end.
#' @export
get_tss_region <- function(gtf, gene.use, upstream_length=1000, downstream_length=500){
  # Checks
  validInput(gtf,'gtf','df')
  validInput(gene.use,'gene.use','character')
  validInput(upstream_length,'upstream_length','numeric')
  validInput(downstream_length,'downstream_length','numeric')
  
  # Preparations
  df <- gtf %>% 
    filter(V3=='gene') %>% 
    mutate(genes = sapply(V9, strsplit, ";| ") %>% sapply("[", 2) %>% unname()) %>%
    filter(genes %in% gene.use)
      
  # Calculate
  res <- df %>%
    apply(1, function(x) {
      if(x[7] == "-") pos = x[5] else pos = x[4] 
      pos %>% 
        as.integer() %>% 
        {c(as.integer(.-upstream_length),as.integer(.+downstream_length))}
    }) %>% 
    t() %>% 
    data.frame() %>%
    mutate(chr = df$V1, genes = df$genes) %>% 
    rename(start = X1, end = X2)

  return(res)
}


#' Use fimo to scan the TSS regions of candidate genes to find binding motifs
#'
#' @param gene_tss TSS region of genes. first column should be gene, second column
#' should be chr, third column should be start, fourth column should be end.
#' generated by \code{\link{get_tss_region}}
#' @param motif motif file, you can choose our bulit-in motif database of
#' 'mus musculus', 'homo sapiens', 'zebrafish' and 'chicken' by 'motif = Tranfac201803_Mm_MotifTFsF',
#' 'motif = Tranfac201803_Hs_MotifTFsF', 'motif = Tranfac201803_Zf_MotifTFsF',
#' 'motif = Tranfac201803_Ch_MotifTFsF' respectively, or you can upload your own
#' motif data base, but the formata use be the same as our built-in motif database.
#' @param refdir character, indicating the path of reference genome. Reference
#' genome can be download from https://hgdownload.soe.ucsc.edu/downloads.html
#' @param fimodir character, indicating path of fimo software,
#' if you have added fimo to the environment variable, just set this argument as 'fimo'
#' @param outputdir1 character, indicating the output path of the shell scripts
#' and sequence of target genes tss regions.(function 'find_motifs_targetgenes'
#' will automatically generate two folders ('fasta' and 'fimo') in the path
#' 'outputdir1', and store sequence of target genes tss regions in the 'fasta'
#' and shell scripts in the 'fimo')
#' @param Motifdir character, indicating the path of meme motif file
#' @param sequencedir character, indicating the path of sequence of target genes
#' tss regions. If it's NULL, this parameter will be paste0(outputdir1,'fasta/')
#' @param select_motif logic, indicating whether to select motifs whose related
#' transcription factors are in genes of gene_tss
#' @param use_nohup logic, indicating whether use nohup to run all fimo scripts simultaneously
#' @return
#' @export
#'
#' @examples
find_motifs_targetgenes <- function(gene_tss, 
                                    motif,
                                    refdir, 
                                    fimodir, 
                                    outputdir, 
                                    Motifdir, 
                                    sequencedir = NULL, 
                                    select_motif = TRUE, 
                                    use_nohup = FALSE, 
                                    verbose = TRUE,
                                    n.cores = 1) {
  # Checks
  validInput(refdir,'refdir','fileexists')
  validInput(Motifdir,'Motifdir','direxists')
  if (!dir.exists(outputdir)) {
    if(verbose) message("Creating output dir")
    dir.create(outputdir)
  }
  if (str_ends(outputdir,'/')==FALSE) warning('the last character of outputdir1 is not "/"')
  if(!is.logical(use_nohup)) stop('parameter use_nohup should be TRUE or FALSE')
  outfasta <- paste0(outputdir,'fasta/')
  outfimo <- paste0(outputdir,'fimo/')
  if (!dir.exists(outfasta)) dir.create(outfasta)
  if (!dir.exists(outfimo)) dir.create(outfimo)
  if (is.null(sequencedir)) sequencedir <- outfasta
  
  # Preparations
  if (select_motif==T) {
    idx <- (motif1$EnsemblID %>% 
             sapply(strsplit, split = ";", fixed = TRUE) %>% 
             sapply(`[[`, 1)) %in% 
      gene_tss$genes
    
    motif1 %<>% filter(idx)
  } else {
    motif1 = motif
  }
  if (verbose) cat("Reading fasta... ")
  fasta <- readBStringSet(refdir, format = "fasta", nrec = -1L,
                                      skip = 0L, seek.first.rec = FALSE,
                                      use.names = TRUE)
  
  if (verbose) cat("done!\nCreating fasta files...\n")
  pbmcmapply(\(gene, chr, start, end) {
    dir.create(paste0(outfimo, gene))
    
    if (end > length(fasta[[chr]])) {
      sequence <- fasta[[chr]][(start + 1):length(fasta[[chr]])]
      name <- paste0(">",chr,":",start,"-",length(fasta[[chr]]))
    } else {
      sequence <- fasta[[chr]][(start + 1):end]
      name <- paste0(">",chr,":",start,"-",end)
    }
    data.frame(name, sequence %>% 
                 as.character() %>% 
                 toupper()) %>% 
      write.table(paste0(outfasta, gene,'.fa'), col.names = FALSE, row.names = FALSE, quote = FALSE)
    
    find_motifs(motif1, step=500, fimodir, paste0(outfimo, gene,'/'), Motifdir, paste0(sequencedir,gene,'.fa')) # Internal function, check outputdirs
  }, gene = gene_tss$genes, chr = gene_tss$chr, start = gene_tss$start, end = gene_tss$end, mc.cores = n.cores)
  
  if (verbose) cat("Creating bash files... ")
  if (file.exists(paste0(outfimo,gene_tss$genes[1],"/Fimo_All.sh"))) out.file <- "/Fimo_All.sh" else out.file <- "/Fimo1.sh"

    sapply(gene_tss$genes, \(gene) {
    if (use_nohup) {
      paste0('nohup sh ',outfimo,gene,out.file," &")
    } else {
      paste0('sh ',outfimo,gene,out.file," ;")
    }
  }) %>% 
    write.table(paste0(outfimo,'fimoall.sh'), quote = FALSE, row.names = FALSE, col.names = FALSE)
  if (verbose) cat("done!")
}

#' Generate regulation database accoding to the fimo result
#'
#' @param motif_dir character, indicating the path of fimo result
#' @param motif motif file, you can choose our bulit-in motif database of
#' 'mus musculus', 'homo sapiens', 'zebrafish' and 'chicken' by 'motif = Tranfac201803_Mm_MotifTFsF',
#' 'motif = Tranfac201803_Hs_MotifTFsF', 'motif = Tranfac201803_Zf_MotifTFsF',
#' 'motif = Tranfac201803_Ch_MotifTFsF' respectively, or you can upload your own
#' motif data base, but the formata use be the same as our built-in motif database.
#' @return
#' @export
#'
#' @examples
generate_fimo_regulation <- function(motif_dir,motif, n.cores=1){
  # Checks
  validInput(motif_dir,'motif_dir','direxists')
  validInput(motif,'motif','df')
  
  # Calculate
  res_final <- plapply(dir(motif_dir), function(i) {
    DirFile1 <- list.files(paste0(motif_dir,i))
    if (!is_empty(DirFile1)) {
      info <- file.info(paste0(motif_dir,i,'/',DirFile1)) %>% 
        `rownames<-`(DirFile1)
        filter(., size > 0, !grepl(c('Fimo_All.sh|Fimo..sh|Fimo...sh'),rownames(.))) %>%
      if (nrow(info)>0) {
        res <- motif %>% 
          filter(Accession %in% (rownames(info) %>% gsub('.txt','', .))) %>%
          apply(1, function(x) strsplit(x[5], ';') %>% sapply("[", 1)) %>% 
          unlist() %>% 
          paste(collapse = ";") %>% 
          data.frame(i)
      }
      return(res)
    }
  }, progress = T, n.cores = n.cores)
  return(res_final)
}

#' Filter regulatory relationships according to binding motifs generated by fimo
#' @description If ATAC-seq data is not available, binding motifs of target genes
#' can be used to filter regulatory relationships.
#' @param fimo_regulation regulation database, you can download it from xxx, or
#' generate it by \code{\link{generate_fimo_regulation}}
#' @param regulatory_relationships  unfiltered regulatory_relationships, generated by
#' \code{\link{get_cor}}
#'
#' @return
#' @export
#'
#' @examples
filter_regulation_fimo <- function(fimo_regulation,regulatory_relationships){
  if (!'TF' %in% colnames(regulatory_relationships)) {
    stop('regulatory_relationships should contain "TF" column')
  }
  if (!'Target' %in% colnames(regulatory_relationships)) {
    stop('regulatory_relationships should contain "Target" column')
  }
  fimo_pair <- apply(fimo_regulation, 1, function(x1){
    TFs <- strsplit(x1[1],';')[[1]]
    Target <- x1[2]
    regulation <- paste(TFs,Target)
    return(regulation)
  })
  fimo_pair <- unlist(fimo_pair)
  regulation_pair <- paste(regulatory_relationships[,1],regulatory_relationships[,4])
  regulation1 <- regulatory_relationships[regulation_pair %in% fimo_pair,]
  return(regulation1)
}
